### PR TITLE
Bind dbaccessor to loopback for k8s

### PR DESCRIPTION
**Pull request by SimonRichardson**
The prior change bootstrapped dqlite to a cloud-local address where possible. This included k8s setup, although it was never required because CAAS allowed HA. Unfortunately, this becomes a problem, if you delete the pod and it becomes rescheduled. When comming back it can't rebind to the address, as it's not the same one.

To fix the problem, we prefer loopback address for CAAS deployments. The node manager now gains that capability to know if the nodes should perfer loopback or not. Then in the dbaccessor worker, we can the ask the node manager for the information.
As the node manager is setup in the manifold for the dbaccessor worker, we can then employ the IAAS vs CAAS manifolds to provide the different functionality. This is a leaf from the 4.0 branch, where we're trying to remove the `if isCAAS`-isms.

## Checklist

- [x] Code style: imports ordered, good names, simple structure, etc
- [x] Comments saying why design decisions were made
- [x] Go unit tests, with comments saying what you're testing

### Microk8s

```sh
$ juju bootstrap microk8s test
```

Ensure it's up and running...

```sh
$ microk8s kubectl logs --namespace=controller-test controller-0 -c api-server -f 
```

Kill the controller pod

```sh
$ microk8s kubectl delete pod controller-0 --namespace=controller-test
```

Make sure it comes back (takes some time to kill the pod)....

```sh
$ microk8s kubectl logs --namespace=controller-test controller-0 -c api-server -f 
```

### Regression test

```sh
$ juju bootstrap lxd test --build-agent
$ juju enable-ha
```

## Links

**Launchpad bug:** https://bugs.launchpad.net/juju/+bug/2051836

**Jira card:** JUJU-5393

### CODE REVIEW

**Review by nvinuesa — APPROVED** [review]
Changes look good to me, but I'd wait for Harry and/or Joe approval

File: cmd/jujud/agent/machine/manifolds.go
Comment (manadart) [inline cmd/jujud/agent/machine/manifolds.go:839]: You could just leave this in the common manifolds and pass a partial call of `NewNodeManager` that wraps over `IsCAASConfig`, which you already have. `proxyConfigUpdater` uses it.

File: worker/dbaccessor/manifold.go
Comment (manadart) [inline worker/dbaccessor/manifold.go:176]: Then you wouldn't need these.

**Review by manadart — APPROVED** [review]
I think I prefer keeping the manifold declaration in the common manifolds due to our wish to back this change out at some future point.

Looks good though; just a few other minor things.

File: cmd/jujud/agent/machine/manifolds.go
Comment (SimonRichardson) [reply cmd/jujud/agent/machine/manifolds.go:839]: I'm not sure I like that either. I'd prefer to lean harder and move proxyConfigUpdater into the different manifolds.

File: worker/dbaccessor/manifold.go
Comment (SimonRichardson) [reply worker/dbaccessor/manifold.go:176]: I like these, it's very explicit that although temporary they're calling it out that there is a different setup between IAAS and CAAS. This is very much, dependency driven.

File: worker/dbaccessor/worker.go
Comment (jack-w-shaw) [inline worker/dbaccessor/worker.go:425]: Why is `ctx` removed here?

Comment (jack-w-shaw) [reply worker/dbaccessor/worker.go:425]: Is it just because it's unused?

File: worker/dbaccessor/worker_test.go
Comment (jack-w-shaw) [inline worker/dbaccessor/worker_test.go:277]: Should this be true?

Comment (manadart) [reply worker/dbaccessor/worker_test.go:277]: Yes. The comments need updating too (my fault) `IsBootstrappedNode` was removed and replaced with `IsLoopbackBound`.

**Pull request merged**
